module_2/gifts/service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,224 @@
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.MultiValueMap;
+
+import com.epam.esm.dto.CertificateDto;
+import com.epam.esm.dto.TagDto;
+import com.epam.esm.exception.ErrorCode;
+import com.epam.esm.exception.NotFoundException;
+import com.epam.esm.repository.CertificateRepository;
+import com.epam.esm.repository.TagRepository;
+import com.epam.esm.repository.model.CertificateModel;
+import com.epam.esm.repository.model.TagModel;
+import com.epam.esm.repository.query_builder.EntityConstant;
+import com.epam.esm.service.CertificateService;
+import com.epam.esm.service.TagService;
+import com.epam.esm.service.converter.CertificateConverter;
+import com.epam.esm.service.converter.TagConverter;
+import com.epam.esm.service.validation.CertificateValidation;
+import com.epam.esm.service.validation.TagValidation;
+import com.epam.esm.service.validation.Util;
+
+/**
+ * 
+ * Contains methods implementation for working mostly with {@code CertificateDto}
+ * entity.
+ *
+ */
+@Service
+public class CertificateServiceImpl implements CertificateService {
+
+	@Autowired
+	private CertificateRepository certificateRepository;
+
+	@Autowired
+	private TagRepository tagRepository;
+
+	@Autowired
+	private CertificateValidation certificateValidation;
+
+	@Autowired
+	private TagValidation tagValidation;
+
+	@Autowired
+	private CertificateConverter certificateConverter;
+
+	@Autowired
+	private TagConverter tagConverter;
+
+	@Autowired
+	private TagService tagService;
+
+	/**
+	 * Creates and saves the passed certificate.
+	 * 
+	 * @param certificateDto the certificate to be saved
+	 * @return saved certificate
+	 */
+	@Override
+	@Transactional
+	public CertificateDto create(CertificateDto certificateDto) {
+		certificateValidation.validateCertificateAllFieldsRequirementsForCreate(certificateDto);
+
+		LocalDateTime now = LocalDateTime.now();
+		certificateDto.setCreateDate(now);
+		certificateDto.setLastUpdateDate(now);
+
+		CertificateModel createdCertificateModel = certificateRepository
+				.create(certificateConverter.convertToModel(certificateDto));
+		CertificateDto createdCertificate = certificateConverter.convertToDto(createdCertificateModel);
+		createdCertificate.setTags(saveCertificateTags(createdCertificate.getId(), certificateDto.getTags()));
+
+		return createdCertificate;
+	}
+
+	/**
+	 * Reads certificate with passed id.
+	 * 
+	 * @param certificateId the id of certificate to be read
+	 * @return certificate with passed id
+	 */
+	@Override
+	public CertificateDto readById(long certificateId) {
+		certificateValidation.checkCertificateExistenceById(certificateId);
+
+		CertificateDto certificateDto = certificateConverter
+				.convertToDto(certificateRepository.readById(certificateId));
+
+		certificateDto.setTags(tagService.readByCertificateId(certificateId));
```

Comment:
Why do you call to sql query instead of use one with join?

Reply:
@maximBortnik, The CertificateModel contains fields only for the certificate database table while CertificateDto contains additional field with the list of tags. That is why the query is called twice: first call to get all fields for the certificate and next call to read tags. Initially I was planned to include list with tags as the field of CertificateModel, but in this case it will reflect not only the certificate table columns in database.

Reply:
To add tags as field into certificate shouldn't effect the certificate table if you has correct many-to-many relationship into database

Reply:
For this request it ok if you has to queries, but see comment bellow. In that case it's a big performance issues